src/main/java/com/openlattice/shuttle/dates/JavaDateTimeHelper.java

```diff
@@ -1,81 +1,123 @@
 package com.openlattice.shuttle.dates;
 
 import com.openlattice.shuttle.util.Cached;
+import com.openlattice.shuttle.util.Constants;
 import org.apache.commons.lang3.StringUtils;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import java.sql.Timestamp;
 import java.time.LocalDate;
 import java.time.LocalDateTime;
 import java.time.LocalTime;
 import java.time.OffsetDateTime;
 import java.time.format.DateTimeFormatter;
 import java.time.format.DateTimeParseException;
+import java.util.Optional;
 import java.util.TimeZone;
 import java.util.concurrent.ExecutionException;
 import java.util.function.BiFunction;
 
 public class JavaDateTimeHelper {
-    private static final Logger logger = LoggerFactory.getLogger(JavaDateTimeHelper.class);
+    private static final Logger logger = LoggerFactory.getLogger( JavaDateTimeHelper.class );
 
-    private final TimeZone      tz;
-    private final String[]      datePatterns;
+    private final Optional<TimeZone> tz;
+    private final String[]           datePatterns;
 
-    public JavaDateTimeHelper(TimeZone tz, String... datePatterns) {
+    public JavaDateTimeHelper( Optional<TimeZone> tz, String... datePatterns ) {
         this.tz = tz;
         this.datePatterns = datePatterns;
     }
 
-    private boolean shouldIgnoreValue(String date) {
-        return date == null || StringUtils.isBlank(date) || date.equals("NULL");
+    private boolean shouldIgnoreValue( String date ) {
+        return date == null || StringUtils.isBlank( date ) || date.equals( "NULL" );
     }
 
-    public <R> R parseWithTwoDigitYearHandling(String date, BiFunction<String, DateTimeFormatter, R> parseFunction,
+    public <R> R parseWithTwoDigitYearHandling(
+            String date, BiFunction<String, DateTimeFormatter, R> parseFunction,
             BiFunction<R, String, R> postParseFunction ) {
-        if (shouldIgnoreValue(date)) {
+        if ( shouldIgnoreValue( date ) ) {
             return null;
         }
-        for ( String datePattern: datePatterns ){
+        for ( String datePattern : datePatterns ) {
             try {
                 DateTimeFormatter formatter = Cached.getDateFormatForString( datePattern );
                 R result = parseFunction.apply( date, formatter );
                 return postParseFunction.apply( result, datePattern );
-            } catch ( DateTimeParseException e) {
-                if ( datePattern.equals(datePatterns[datePatterns.length - 1]) ) {
-                    logger.error("Unable to parse date {}, please see debug log for additional information: {}.", date,e);
+            } catch ( DateTimeParseException e ) {
+                if ( datePattern.equals( datePatterns[ datePatterns.length - 1 ] ) ) {
```

Comment:
This is a bit confusing and can potentially do bad things if someone messes up and includes the same date twice. I'd find it a bit more straightforward to do:
```
for ( int i = 0; i < datePatterns.length; i++ ) {

  try {
    DateTimeFormatter formatter = Cached.getDateFormatForString( datePatterns[ i ] );
    R result = parseFunction.apply( date, formatter );
    return postParseFunction.apply( result, datePattern );
  } catch ( DateTimeParseException e ) {
    if (i == datePatterns.length - 1) {
      logger.error( "Unable to parse date {}, please see debug log for additional information: {}.",
                            date,
                            e );
    }
  }
}
```

Further, it looks like there's a lot of code repetition in this class following this pattern. Whenever you see this, that's a good sign to look for ways to generify the repeated logic, which makes maintaining the code way easier and helps prevent things from breaking in the future in the event that some component needs to be updated.

For example, `parseWithResultType` could be simplified as this: 
```
    public <R> R parseWithResultType( String parseableString, BiFunction<String, DateTimeFormatter, R> parseFunction ) {
        return parseWithTwoDigitYearHandling( parseableString, parseFunction, (r, str) -> r );
    }
```

Reply:
First comment: good catch.
For the second, I made the change but am not suuuuper-confident.  I don't think this was there before, so there must have been a reason why this might have been added.

Reply:
Yeah I'm not sure why it was added, but if you look at the code, the two functions are identical except that `parseWithTwoDigitYearHandling` includes a post-processing step based on a function that's passed in. If you call it with a no-op function, they do the exact same thing.